WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -251,27 +264,30 @@ public IEnumerable<Money> Decompose(IEnumerable<Money> myInputCoinEffectiveValue
 		// This can happen when smallest denom is larger than the input sum.
 		if (naiveSet.Count == 0)
 		{
-			naiveSet.Add(remaining);
+			naiveSet.Add(Output.FromAmount(remaining, ChangeScriptType, FeeRate));
 		}
 
 		HashCode hash = new();
-		foreach (var item in naiveSet.OrderBy(x => x))
+		foreach (var item in naiveSet.OrderBy(x => x.EffectiveCost))
 		{
-			hash.Add(item);
+			hash.Add(item.Amount);
 		}
 
 		setCandidates.Add(
 			hash.ToHashCode(), // Create hash to ensure uniqueness.
-			(naiveSet, loss + (ulong)naiveSet.Count * (OutputFee + InputFee))); // The cost is the remaining + output cost + input cost.
+			(naiveSet, loss + CalculateCostMetrics(naiveSet)));
 
 		// Create many decompositions for optimization.
-		var stdDenoms = denoms.Where(x => x <= myInputSum).Select(x => (long)x).ToArray();
-		var maxNumberOfOutputsAllowed = Math.Min(AvailableVsize / OutputSize, 8);
+		var stdDenoms = denoms.Where(x => x.Amount <= myInputSum).Select(d => d.EffectiveCost.Satoshi).ToArray();
+		var smallestScriptType = Math.Min(ScriptType.P2WPKH.EstimateOutputVsize(), ScriptType.Taproot.EstimateOutputVsize());
+		var maxNumberOfOutputsAllowed = Math.Min(AvailableVsize / smallestScriptType, 8); // The absolute max possible with the smallest script type.
```

Comment:
That is the minimum. We are not going for safety here, we cannot - because we do not know the script types. We select more and later filter out the invalid ones. 

https://github.com/molnard/WalletWasabi/blob/taprootdecomp/WalletWasabi/WabiSabi/Client/AmountDecomposer.cs#L308